### PR TITLE
Updated plugin list format

### PR DESCRIPTION
**Pull request by mathiascode**
One potential solution for issues where an ini parser removes duplicate keys when the plugin list is loaded.

```
[Plugins]
WorldEdit=1
Core=1
ChatLog=1
ProtectionAreas=0
```

1 enables a plugin, 0 disables it. Commenting out a line can also be used to disable a plugin. I also thought about having a single line listing all enabled plugins with commas, but it would be harder to quickly disable a plugin.

### CODE REVIEW

**Comment by madmaxoft** [comment]
Simple, working, and has an upgrade path. /me likes.

Perhaps a comment should be added to the upgrade path, as to when this was upgraded, so that we can remove it after some time.

File: src/Bindings/PluginManager.cpp [hunk 1]
```diff
@@ -1733,22 +1734,35 @@ AStringVector cPluginManager::GetFoldersToLoad(cSettingsRepositoryInterface & a_
...
 	{
 		AString ValueName = NameValue.first;
 		if (ValueName.compare("Plugin") == 0)
 		{
 			AString PluginFile = NameValue.second;
 			if (!PluginFile.empty())
 			{
-				res.push_back(PluginFile);
+				a_Settings.DeleteValue("Plugins", ValueName);
+				a_Settings.SetValue("Plugins", PluginFile, "1");
```
Comment (peterbell10) [inline src/Bindings/PluginManager.cpp]: If I have a plugin called `Plugin` then I might write `Plugin=1`. This would then replace it with `1=1`.

Comment (mathiascode) [reply src/Bindings/PluginManager.cpp]: This is why code reviews are great.

File: src/Bindings/PluginManager.cpp [hunk 2]
```diff
@@ -1733,22 +1734,40 @@ AStringVector cPluginManager::GetFoldersToLoad(cSettingsRepositoryInterface & a_
 		InsertDefaultPlugins(a_Settings);
 	}
 
-	// Get the list of plugins to load:
 	AStringVector res;
-	auto Values = a_Settings.GetValues("Plugins");
-	for (auto NameValue : Values)
+
+	// Get the old format plugin list, and migrate it.
+	// Upgrade path added on 27/03/2020
```
Comment (madmaxoft) [inline src/Bindings/PluginManager.cpp]: Please use the one and only ISO date format: 2020-03-27

File: src/Bindings/PluginManager.cpp [hunk 3]
```diff
@@ -1733,22 +1734,40 @@ AStringVector cPluginManager::GetFoldersToLoad(cSettingsRepositoryInterface & a_
...
 			AString PluginFile = NameValue.second;
-			if (!PluginFile.empty())
+			if (
+				!PluginFile.empty() &&
+				(PluginFile != "0") &&
+				(PluginFile != "1")
+			)
 			{
-				res.push_back(PluginFile);
+				a_Settings.DeleteValue("Plugins", ValueName);
```
Comment (madmaxoft) [inline src/Bindings/PluginManager.cpp]: Is it possible that this deletion can become confused about which of the `Plugin=` lines to remove?

File: src/Bindings/PluginManager.cpp (hunk 2 above)
Comment (mathiascode) [reply src/Bindings/PluginManager.cpp]: The chosen one

File: src/Bindings/PluginManager.cpp (hunk 3 above)
Comment (mathiascode) [reply src/Bindings/PluginManager.cpp]: DeleteValue should remove the first value it finds.

**Comment by madmaxoft** [comment]
OH NO, we completely missed one aspect. The old format specified not only the plugins to load, but also their load order. The new format doesn't specify the load order. Some plugins might have depended on the order.

**Comment by bearbin** [comment]
I'm not sure I understand - with this PR plugins will still be loaded in the order they are in the INI file, won't they?

**Comment by madmaxoft** [comment]
They are? Still that's a non-standard INI file usage, as the key order in the INI file is usually not relevant.
Okay, I'll believe this then and won't mention it again.

**Comment by tigerw** [comment]
> OH NO

this is too funny XD

I am amused at the 9-day delay the epiphany took. Was it one of those "suddenly think of an embarrassing event five years ago whilst lying awake at night" moments?

**Comment by bearbin** [comment]
OTOH, we did have the opportunity to add proper ordering not relying on ordering in the file, if we allow any number > 0 as a value for the plugin key, then the list of plugins to load could be sorted by this with higher-numbered plugins being loaded after lower numbered ones. We could still do that, it would be backwards compatible.

**Comment by madmaxoft** [comment]
I'd say we should do that, @bearbin , great idea. Make it consistent.

@tigerw yes, one of those, just when I was adding my new plugin to the list on my dev server.

**Comment by mathiascode** [comment]
Have fun writing documentation for the average user. :P

**Pull request merged**